stumpy/core.py

```diff
@@ -679,15 +706,30 @@ def rolling_nanstd(a, w):
     w : numpy.ndarray
         The rolling window size
 
+    welford : bool, default False
+        When False (default), the computation is parallelized and the std of
+        each subsequence is calculated on its own. When `welford==True`, the
+        welford method is used to reduce the computing time at the cost of slight
+        loss of precision
+
     Returns
     -------
-    output : numpy.ndarray
-        Rolling window nanstd.
+    out : numpy.ndarray
+        Rolling window nanstd. When `a` is 1D, `out[i]` is the std of `a[i, i + w]`.
+        When `a` is 2D, `out[i, j]` is the std of `a[i, j : j + w]`.
     """
+    if a.ndim > 2:  # pragma nocover
```

Comment:
Wouldn't this still work for for 3D or 4D or ND arrays?

Reply:
I think it works but I do not know if it would be meaningful for this function since `np.nanstd(..)` will be the nanstd of flattened array.

Reply:
In that case, shouldn't we change `_rolling_nanstd_1d` to:

```
out[i] = np.nanstd(a[..., i : i + w], axis=a.ndim)
```

Reply:
Danggg, `numba` doesn't support `axis` argument for `np.nanstd`

Reply:
This is going to get dirty as the behavior is going to be different form the other rolling functions

Reply:
As mentioned in other functions, this function should be compared against the output of its numpy equivalent of:

```
np.nanstd(rolling_window(a[..., :], w), axis=a.ndim)
```

Reply:
> Danggg, `numba` doesn't support `axis` argument for `np.nanstd`. Hmmmm

Yeah...it only supports the first argument.


> As mentioned in other functions, this `rolling_nanstd` function should be compared against the output of its numpy equivalent of:
> 
> ```
> np.nanstd(rolling_window(a[..., start:stop], w), axis=a.ndim)
> ```
> 
> This should be tested for multiple dimensions

I have no solution for it as of now. I am going to think about it...

Reply:
Okay, let's not worry about generalizing it. However, is it correct for the 2D case? Something feels wrong

Reply:
> Okay, let's not worry about generalizing it

Sure.

> However, is it correct for the 2D case? Something feels wrong

I just created the following naive function. 
```
def naive_std(T, m):
    return np.nanstd(core.rolling_window(T, m), axis=T.ndim)
```
I compared `ref` with `comp` for 2D array `T` using `npt.assert_almost_equal`. It raises no error and everything is good. 


---

personally speaking, it was hard for me to quickly understand how `numpy.apply_along_axis` works. Because, for a 2D input T, where each row is a time series, `numpy.apply_along_axis(...., axis=1, ...)` confused me a little bit. After reading the examples provided in the numpy doc, I can better understand how this function works.


Side note: 
The [numpy doc](https://numpy.org/doc/stable/reference/generated/numpy.apply_along_axis.html) says:
> Parameters
func1dfunction (M,) -> (Nj…)
This function should accept 1-D arrays. It is applied to 1-D slices of arr along the specified axis.

However, when I tried:
```
def my_func(a):
    return (a[0] + a[-1]) * 0.5

A = np.random.randint(1, 100, size=(3, 4, 5))
np.apply_along_axis(my_func, axis=0, arr=A)
```

It works. I think the slice of A along the specified axis is not 1D in this example.

Reply:
@seanlaw 
**So, I have been working on this, and I found the following:**

It seems that the new function works correctly for `a` with `ndim > 2`. I do not know why/how :) I am trying to understand how `core._rolling_nanstd_1d` works when input is not 1D and no axis is specified in `np.nanstd(..)`!!!


```
Example

# input
a = np.random.rand(4 * 3 * 17).reshape(4, 3, 17)
w = 5

# naive
start = 0
stop = 18
A = core.rolling_window(a[..., start:stop], w) # shape:  (4, 3, 13, 5)
ref_std = np.nanstd(A, axis=a.ndim) # shape: (4, 3, 13)

 
# our new func when `welford==False`
axis = a.ndim - 1
comp_std = np.apply_along_axis(
            lambda a_row, w: core._rolling_nanstd_1d(a_row, w), axis=axis, arr=a, w=w
        ) 

# comp_std.shape: (4, 3, 13)

>>> npt.assert_almost_equal(ref_std, comp_std)
True
```

Reply:
I think this is how `numpy.apply_along_axis(func1d, ...)` works: 

Let's say `arr` has shape of `(2, 3, 4)` and `axis=1`, then it applies `func1d` on all of the following slices:
```
arr[0, :, 0]
arr[0, :, 1]
arr[0, :, 2]
arr[0, :, 3]
arr[0, :, 4]
...
arr[1, :, 1]
arr[1, :, 2]
arr[1, :, 3]
arr[1, :, 4]
```

So, we do not need to specify `axis` for `np.nanstd`.

Reply:
Ahhh, interesting! I just took a look at it and I am beginning to understand it better. So, each function must still handle the "rolling" calculation along a 1D array on its own but the `np.apply_along_axis` slices everything up for us. And we still benefit from `parallel=True` and `prange` since the function still needs to roll along the 1D array slice.

I guess we got lucky! This is great

Reply:
> Should we remove the `ValueError` then?

Yeah...I will fix it now.